blockingconcurrentqueue.h

```diff
@@ -654,6 +697,19 @@ class BlockingConcurrentQueue
 			continue;
 		}
 	}
+
+	template<typename U>
+	inline bool wait_dequeue_with_timeout(int64_t timeout_usecs, U& item)
+	{
+		// 0us is still a timeout, hence >=
+		if (timeout_usecs >= 0 && !sema->wait_for_usecs(timeout_usecs)) {
```

Comment:
There's a bug here; the semaphore must *always* be decremented before removing something from the queue, even if the timeout is infinite or negative.